src/lap_stats/lap_stats.c

```diff
@@ -391,67 +387,75 @@ static void setupGeoTriggers(const TrackConfig *tc, const Track *track) {
-      const TrackConfig *trackConfig = &(config->TrackConfigs);
-      const Track *track = &trackConfig->track;
-      if (trackConfig->auto_detect){
-    	  track = auto_configure_track(track, gp);
-      }
-      else{
-    	  pr_info("track: using fixed config");
-      }
-      set_active_track(track);
-
-      startFinishEnabled = isStartFinishEnabled(track);
-      sectorEnabled = isSectorTrackingEnabled(track);
-
-      lc_reset();
-      lc_setup(track, targetRadius);
-      setupGeoTriggers(trackConfig, track);
-      g_configured = 1;
-   }
-
-   if (!startFinishEnabled) return;
-
-   // Process data fields first.
-   updateGeoTrigger(&startGeoTrigger, gp);
-   updateGeoTrigger(&finishGeoTrigger, gp);
-   update_elapsed_time(gpsSnapshot);
-   update_distance(gpsSnapshot);
-   addGpsSample(gpsSnapshot);
-
-   /*
+        g_finish_geo_trigger = createGeoTrigger(&gc);
+}
+
+
+static void lapstats_location_updated(const GpsSnapshot *gps_snapshot) {
+    const LoggerConfig *config = getWorkingLoggerConfig();
+    const GeoPoint *gp = &gps_snapshot->sample.point;
+    const float target_radius = degrees_to_meters(config->TrackConfigs.radius);
+
+    if (!g_start_finish_enabled) return;
+
+    // Process data fields first.
+    updateGeoTrigger(&g_start_geo_trigger, gp);
+    updateGeoTrigger(&g_finish_geo_trigger, gp);
+    update_elapsed_time(gps_snapshot);
+    update_distance(gps_snapshot);
+    addGpsSample(gps_snapshot);
+
+    /*
     * Now process the sector, finish and start logic in that order.
     * Each processing can invoke their respective event if the logic
     * agrees its time.
     */
-   //Skip processing sectors bits if we haven't defined any... duh!
-   if (sectorEnabled) {
-           processSectorLogic(gpsSnapshot, g_activeTrack, targetRadius);
-   }
-   processFinishLogic(gpsSnapshot, g_activeTrack, targetRadius);
-   processStartLogic(gpsSnapshot, g_activeTrack, targetRadius);
-}
-
-void lapStats_processUpdate(const GpsSnapshot *gpsSnapshot) {
+    //Skip processing sectors bits if we haven't defined any... duh!
+    if (g_sector_enabled) {
+        processSectorLogic(gps_snapshot, g_activeTrack, target_radius);
+    }
+    processFinishLogic(gps_snapshot, g_activeTrack, target_radius);
+    processStartLogic(gps_snapshot, g_activeTrack, target_radius);
+}
+
+static void lapstats_setup(const GpsSnapshot *gps_snapshot){
+    const LoggerConfig *config = getWorkingLoggerConfig();
+    const float target_radius = degrees_to_meters(config->TrackConfigs.radius);
+    const GeoPoint *gp = &gps_snapshot->sample.point;
+
+    const Track *track = NULL;
+    const TrackConfig *trackConfig = &(config->TrackConfigs);
+    if (trackConfig->auto_detect){
+        track = auto_configure_track(NULL, gp);
+    }
+    else{
+        track = &trackConfig->track;
+        g_track_status = TRACK_STATUS_FIXED_CONFIG;
+        pr_info("track: using fixed config");
+    }
+    set_active_track(track);
+    if (track){
```

Comment:
IMHO if it gets too confusing then its because your method is doing too much and you need to break it up into something smaller.  Multiple exits are a wonderful feature and make the flow of code much easier to read and save on tons of indentation and potentially complex if/if else/else statements.